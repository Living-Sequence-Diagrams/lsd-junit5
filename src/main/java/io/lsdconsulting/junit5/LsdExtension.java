package io.lsdconsulting.junit5;

import com.lsd.core.IdGenerator;
import com.lsd.core.LsdContext;
import com.lsd.core.properties.LsdProperties;
import j2html.tags.UnescapedText;
import org.junit.jupiter.api.extension.AfterAllCallback;
import org.junit.jupiter.api.extension.AfterTestExecutionCallback;
import org.junit.jupiter.api.extension.ExtensionContext;
import org.junit.jupiter.api.extension.TestWatcher;
import org.junit.platform.commons.util.ExceptionUtils;

import java.lang.annotation.Annotation;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Optional;
import java.util.function.Consumer;

import static com.lsd.core.domain.Status.*;
import static com.lsd.core.report.PopupContent.popupDiv;
import static j2html.TagCreator.*;
import static java.util.Arrays.stream;
import static java.util.function.Predicate.not;
import static org.junit.platform.commons.util.StringUtils.isBlank;

/**
 * Junit 5 extension to create LSD reports based on test results.
 * (Not that this extension does not capture any events for the LSD diagrams)
 */
public class LsdExtension implements TestWatcher, AfterTestExecutionCallback, AfterAllCallback {

    private final LsdContext lsdContext = LsdContext.getInstance();
    private final IdGenerator idGenerator = lsdContext.getIdGenerator();
    private final boolean hideStacktrace = LsdProperties.getBoolean("lsd.junit5.hideStacktrace");

    @Override
    public void testSuccessful(ExtensionContext context) {
        lsdContext.completeScenario(prefixParentDisplayName(context), p(
                h4().with(new UnescapedText("&#127808; Test Passed")).withClass("success")
        ).render(), SUCCESS);
    }

    @Override
    public void testDisabled(ExtensionContext context, Optional<String> reason) {
        lsdContext.completeScenario(prefixParentDisplayName(context), p(
                h4("Test Disabled").withClass("warn")
        ).render(), FAILURE);
    }

    @Override
    public void testAborted(ExtensionContext context, Throwable cause) {
        var description = createErrorDescription(cause, "Test Aborted!");
        lsdContext.completeScenario(prefixParentDisplayName(context), description, FAILURE);
    }

    @Override
    public void testFailed(ExtensionContext context, Throwable cause) {
        var description = createErrorDescription(cause, "&#10060; Failed!");
        lsdContext.completeScenario(prefixParentDisplayName(context), description, ERROR);
    }

    @Override
    public void afterTestExecution(ExtensionContext context) {
        Object testInstance = context.getRequiredTestInstance();
        additionalProcessing(testInstance, LsdPostTestProcessing.class);
    }

    @Override
    public void afterAll(ExtensionContext context) {
        if (isNested(context)) {
            return;
        }
        lsdContext.completeReport(context.getDisplayName());
        lsdContext.createIndex();
    }

    private boolean isNested(ExtensionContext context) {
        return context.getParent()
                .map(ExtensionContext::getParent)
                .map(Optional::isPresent)
                .orElse(false);
    }

    private String prefixParentDisplayName(ExtensionContext context) {
        var parent = context.getParent();
        if (parent.isPresent()) {
            var parentDisplayName = prefixParentDisplayName(parent.get());
            var separator = isBlank(parentDisplayName) ? "" : ": ";
            return parentDisplayName + separator + context.getDisplayName();
        }
        return "";
    }

    private String createErrorDescription(Throwable cause, String header) {
        var contentId = idGenerator.next();
        String exceptionMessage = extractMessage(cause);
        return p(
                h4().with(new UnescapedText(header)).withClass("error"),
                a().withHref("#" + contentId).withText(exceptionMessage),
                popupDiv(contentId, "Stacktrace", readStackTrace(cause))
        ).render();
    }

    private void additionalProcessing(final Object instance, Class<? extends Annotation> annotation) {
        Class<?> klass = instance.getClass();
        while (klass != Object.class) {
            stream(klass.getDeclaredMethods())
                    .filter(method -> method.isAnnotationPresent(annotation))
                    .forEach(invokeMethodOn(instance));
            klass = klass.getSuperclass();
        }
    }

    private String extractMessage(Throwable cause) {
        return Optional.ofNullable(cause)
                .map(Throwable::getMessage)
                .orElse("");
    }

    private String readStackTrace(Throwable cause) {
        return Optional.ofNullable(cause)
                .filter(not(throwable -> hideStacktrace))
                .map(ExceptionUtils::readStackTrace)
                .orElse("[Displaying the stacktrace was disabled or no cause was provided]");
    }

    private Consumer<Method> invokeMethodOn(Object instance) {
        return method -> {
            try {
                method.setAccessible(true);
                method.invoke(instance);
            } catch (IllegalAccessException | InvocationTargetException e) {
                throw new RuntimeException(e);
            }
        };
    }
}