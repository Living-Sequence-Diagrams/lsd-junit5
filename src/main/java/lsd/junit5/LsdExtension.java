package lsd.junit5;

import com.lsd.IdGenerator;
import com.lsd.LsdContext;
import com.lsd.properties.LsdProperties;
import j2html.tags.UnescapedText;
import org.junit.jupiter.api.extension.AfterAllCallback;
import org.junit.jupiter.api.extension.AfterTestExecutionCallback;
import org.junit.jupiter.api.extension.ExtensionContext;
import org.junit.jupiter.api.extension.TestWatcher;
import org.junit.platform.commons.util.ExceptionUtils;

import java.lang.annotation.Annotation;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Optional;
import java.util.function.Consumer;

import static com.lsd.OutcomeStatus.*;
import static com.lsd.report.model.PopupContent.popupDiv;
import static j2html.TagCreator.*;
import static java.util.Arrays.stream;
import static java.util.function.Predicate.not;
import static org.junit.platform.commons.util.StringUtils.isBlank;

public class LsdExtension implements TestWatcher, AfterTestExecutionCallback, AfterAllCallback {

    private final LsdContext lsdContext = LsdContext.getInstance();
    private final IdGenerator idGenerator = lsdContext.getIdGenerator();
    private final boolean hideStacktrace = LsdProperties.getBoolean("lsd.junit5.hideStacktrace");

    @Override
    public void testSuccessful(ExtensionContext context) {
        lsdContext.completeScenario(prefixParentDisplayName(context), p(
                h4().with(new UnescapedText("&#127808; Test Passed")).withClass(SUCCESS.getCssClass())
        ).render(), SUCCESS);
    }

    @Override
    public void testDisabled(ExtensionContext context, Optional<String> reason) {
        lsdContext.completeScenario(prefixParentDisplayName(context), p(
                h4("Test Disabled").withClass(WARN.getCssClass())
        ).render(), WARN);
    }

    @Override
    public void testAborted(ExtensionContext context, Throwable cause) {
        var description = createErrorDescription(cause, "Test Aborted!");
        lsdContext.completeScenario(prefixParentDisplayName(context), description, WARN);
    }

    @Override
    public void testFailed(ExtensionContext context, Throwable cause) {
        var description = createErrorDescription(cause, "&#10060; Failed!");
        lsdContext.completeScenario(prefixParentDisplayName(context), description, ERROR);
    }

    @Override
    public void afterTestExecution(ExtensionContext context) {
        Object testInstance = context.getRequiredTestInstance();
        additionalProcessing(testInstance, LsdPostTestProcessing.class);
    }

    @Override
    public void afterAll(ExtensionContext context) {
        if (notNestedTest(context)) {
            lsdContext.completeReport(context.getDisplayName());
            lsdContext.createIndex();
        }
    }

    private boolean notNestedTest(ExtensionContext context) {
        return !context.getParent()
                .map(ExtensionContext::getParent)
                .map(Optional::isPresent)
                .orElse(true);
    }

    private String prefixParentDisplayName(ExtensionContext context) {
        var parent = context.getParent();
        if (parent.isPresent()) {
            var parentDisplayName = prefixParentDisplayName(parent.get());
            var separator = isBlank(parentDisplayName) ? "" : ": ";
            return parentDisplayName + separator + context.getDisplayName();
        }
        return "";
    }

    private String createErrorDescription(Throwable cause, String header) {
        var contentId = idGenerator.next();
        String exceptionMessage = extractMessage(cause);
        return p(
                h4().with(new UnescapedText(header)).withClass(ERROR.getCssClass()),
                a().withHref("#" + contentId).withText(exceptionMessage),
                popupDiv(contentId, "Stacktrace", readStackTrace(cause))
        ).render();
    }

    private void additionalProcessing(final Object instance, Class<? extends Annotation> annotation) {
        Class<?> klass = instance.getClass();
        while (klass != Object.class) {
            stream(klass.getDeclaredMethods())
                    .filter(method -> method.isAnnotationPresent(annotation))
                    .forEach(invokeMethodOn(instance));
            klass = klass.getSuperclass();
        }
    }

    private String extractMessage(Throwable cause) {
        return Optional.ofNullable(cause)
                .map(Throwable::getMessage)
                .orElse("");
    }

    private String readStackTrace(Throwable cause) {
        return Optional.ofNullable(cause)
                .filter(not(throwable -> hideStacktrace))
                .map(ExceptionUtils::readStackTrace)
                .orElse("[Displaying the stacktrace was disabled or no cause was provided]");
    }

    private Consumer<Method> invokeMethodOn(Object instance) {
        return method -> {
            try {
                method.setAccessible(true);
                method.invoke(instance);
            } catch (IllegalAccessException | InvocationTargetException e) {
                throw new RuntimeException(e);
            }
        };
    }
}