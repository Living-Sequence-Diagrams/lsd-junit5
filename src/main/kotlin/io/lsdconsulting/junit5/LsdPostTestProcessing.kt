package io.lsdconsulting.junit5

/**
 * Test that use the `LsdExtension` to generate LSD reports can use this annotation to flag a
 * method (can be private but must take no arguments) that should be invoked after each Test has run but before the
 * sequence diagram is generated.
 *
 *
 * This may be useful if it is necessary to add additional interactions or interesting givens to the TestState instance
 * before the sequence diagrams are generated.
 */
@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
annotation class LsdPostTestProcessing  