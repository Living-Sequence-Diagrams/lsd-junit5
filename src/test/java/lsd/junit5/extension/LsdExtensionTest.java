package lsd.junit5.extension;

import com.lsd.LsdContext;
import com.lsd.events.Message;
import lsd.junit5.LsdExtension;
import lsd.junit5.LsdPostTestProcessing;
import lsd.junit5.LsdPreTestProcessing;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtensionContext;
import org.mockito.Mockito;

import java.util.Optional;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class LsdExtensionTest {
    private ExtensionContext mockContext = mock(ExtensionContext.class, Mockito.RETURNS_DEEP_STUBS);

    private LsdContext lsdContext = LsdContext.getInstance();
    private LsdExtension lsdExtension = new LsdExtension();

    @Test
    void createsScenarioWhenInformedOfTestFailure() {
        when(mockContext.getRequiredTestInstance()).thenReturn(this);
        when(mockContext.getDisplayName())
                .thenReturn("The failed test name")
                .thenReturn("The successful test name")
                .thenReturn("The aborted test name")
                .thenReturn("The disabled test name");

        lsdExtension.testFailed(mockContext, new RuntimeException("Failed for testing purposes"));
        lsdExtension.testSuccessful(mockContext);
        lsdExtension.testAborted(mockContext, new RuntimeException("Aborted for testing reasons"));
        lsdExtension.testDisabled(mockContext, Optional.of("Aborted reason"));

        Approvals.verify(lsdContext.completeReport("LsdExtension Test Report").toFile());
    }

    @LsdPreTestProcessing
    public void preTestProcessing() {
        lsdContext.capture(Message.builder()
                .id("abc123")
                .from("LsdExtension")
                .to("LsdExtensionTest")
                .data("preTestProcessing()")
                .label("pre test processing")
                .build());
    }

    @LsdPostTestProcessing
    public void postTestProcessing() {
        lsdContext.capture(Message.builder()
                .id("abc456")
                .from("LsdExtension")
                .to("LsdExtensionTest")
                .data("postTestProcessing()")
                .label("post test processing")
                .build());
    }
}
