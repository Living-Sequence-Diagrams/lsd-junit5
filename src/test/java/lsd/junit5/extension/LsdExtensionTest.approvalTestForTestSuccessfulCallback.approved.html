<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xmlns="http://www.w3.org/1999/html" xml:lang="en" lang="en">
<head>
    <title>Successful Test Report</title>
    <link rel="stylesheet" href="style.css">
</head>
<body>
<main role="main">
    <h1 class="logo">LSD</h1>
    <h1>Successful Test Report</h1>
    <div class="contents">
        <h3>Contents:</h3>
        <ul class="contents">
                        <li class="success"><a href="#1">The test name</a></li>
                    </ul>
    </div>
    
    <article id="1">
        <h2 class="success">The test name</h2>
        <section class="description success">
            <h3 class="success">Description</h3>
            <p><h4 class="success">&#127808; Test Passed</h4></p>        </section>
                <section class="sequence diagram svg success">
            <h3 class="success">Sequence Diagram</h3>
            <pre><?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" contentScriptType="application/ecmascript" contentStyleType="text/css" height="12px" preserveAspectRatio="none" style="width:12px;height:12px;background:#FFFFFF;" version="1.1" viewBox="0 0 12 12" width="12px" zoomAndPan="magnify">
  <defs />
  <g />
</svg>

</pre>
            <div onclick="location.href='#!';" id="2" class="overlay">
                <div class="popup" onclick="event.stopPropagation();">
                    <h2>PlantUml markup</h2>
                    <a class="close" href="#!">&times;</a>
                    <div class="content">
                        <pre><code>@startuml

!theme plain
hide unlinked




@enduml</code></pre>
                    </div>
                </div>
            </div>
            <sub><a href="#2">[source]</a></sub>
            <sub><a href="#3">[component diagram]</a></sub>
        </section>
        <!-- Popup for component diagram -->
        <div onclick="location.href='#!';" id="3" class="overlay">
            <div class="popup" onclick="event.stopPropagation();">
                <h2>Component Diagram</h2>
                <a class="close" href="#!">&times;</a>
                <div class="content">
                    <sup>[Requires Graphviz to be installed]</sup>
                    <pre><?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" contentScriptType="application/ecmascript" contentStyleType="text/css" height="12px" preserveAspectRatio="none" style="width:12px;height:12px;background:#FFFFFF;" version="1.1" viewBox="0 0 12 12" width="12px" zoomAndPan="magnify">
  <defs />
  <g />
</svg>

</pre>
                </div>
                <sub><a href="#uml3">[source]</a></sub>
            </div>
        </div>
        <!-- Popup for component diagram UML via component diagram popup -->
        <div onclick="location.href='#!';" id="uml3" class="overlay">
            <div class="popup" onclick="event.stopPropagation();">
                <h2>PlantUml markup</h2>
                <a class="close" href="#3">&times;</a><!--show previous popup on close-->
                <div class="content">
                    <pre><code>@startuml

!theme plain


hide @unlinked
@enduml</code></pre>
                </div>
            </div>
        </div>
        <!-- Section containing data for sequence diagram popups -->
            </article>
    </main>
</body>
<footer>
    <link rel="stylesheet"
          href="https://cdn.jsdelivr.net/gh/highlightjs/cdn-release@11.0.0/build/styles/default.min.css">
    <script src="https://cdn.jsdelivr.net/gh/highlightjs/cdn-release@11.0.0/build/highlight.min.js"></script>
    <script>hljs.highlightAll();</script>

    <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js"></script>
    <script>
        var options = { 
                "element": "span", 
                "className": "highlight", 
                "separateWordSearch": false,
                "acrossElements": true,
                "accuracy": { 
                    "value": "exactly", "limiters": [ ",", ".", ";", ":", "-", "(", ")"] 
                } 
        }
                                    <!-- Remove highlights in svg (causes text to disappear) -->
        instance = new Mark(document.querySelectorAll("svg"));
        instance.unmark();
    </script>
</footer>
</html>
